packages/router/src/router.tsx

```diff
@@ -1,41 +1,221 @@
+}
 
 export function useLocation() {
 	return useContext( RoutesContext );
 }
 
 export function useHistory() {
-	return useContext( HistoryContext );
+	const { pathArg, beforeNavigate } = useContext( ConfigContext );
+	return useMemo(
+		() => ( {
+			navigate( rawPath: string, options: NavigationOptions = {} ) {
+				const query = getQueryArgs( rawPath );
+				const path = getPath( 'http://domain.com/' + rawPath ) ?? '';
+				const performPush = () => {
+					const result = beforeNavigate
+						? beforeNavigate( { path, query } )
+						: { path, query };
+					return history.push(
+						{
+							search: buildQueryString( {
+								[ pathArg ]: result.path,
+								...result.query,
+							} ),
+						},
+						options.state
+					);
+				};
+
+				/*
+				 * Skip transition in mobile, otherwise it crashes the browser.
+				 * See: https://github.com/WordPress/gutenberg/pull/63002.
+				 */
+				const isMediumOrBigger =
+					window.matchMedia( '(min-width: 782px)' ).matches;
+				if (
+					! isMediumOrBigger ||
+					// @ts-expect-error
+					! document.startViewTransition ||
+					! options.transition
+				) {
+					return performPush();
+				}
+				document.documentElement.classList.add( options.transition );
+				// @ts-expect-error
+				const transition = document.startViewTransition( () =>
+					performPush()
+				);
+				transition.finished.finally( () => {
+					document.documentElement.classList.remove(
+						options.transition ?? ''
+					);
+				} );
+			},
+		} ),
+		[ pathArg, beforeNavigate ]
+	);
 }
 
-export function RouterProvider( { children }: { children: React.ReactNode } ) {
+export default function useMatch(
+	location: LocationWithQuery,
+	routes: Route[],
+	pathArg: string
+): Match {
+	const { query: rawQuery = {} } = location;
+
+	return useMemo( () => {
+		const matcher = new RouteRecognizer();
+		routes.forEach( ( route ) => {
+			matcher.add( [ { path: route.path, handler: route } ], {
+				as: route.name,
+			} );
+		} );
+		const { [ pathArg ]: path = '/', ...query } = rawQuery;
+		const result = matcher.recognize( path )?.[ 0 ];
+		if ( ! result ) {
+			return {
+				name: '404',
+				path: addQueryArgs( path, query ),
+				areas: {},
+				widths: {},
+				query,
+				params: {},
+			};
+		}
+
+		const matchedRoute = result.handler as Route;
+		const resolveFunctions = ( record: Record< string, any > = {} ) => {
+			return Object.fromEntries(
+				Object.entries( record ).map( ( [ key, value ] ) => {
+					if ( typeof value === 'function' ) {
+						return [
+							key,
+							value( { query, params: result.params } ),
+						];
+					}
+					return [ key, value ];
+				} )
+			);
+		};
+		return {
+			name: matchedRoute.name,
+			areas: resolveFunctions( matchedRoute.areas ),
+			widths: resolveFunctions( matchedRoute.widths ),
+			params: result.params,
+			query,
+			path: addQueryArgs( path, query ),
+		};
+	}, [ routes, rawQuery, pathArg ] );
+}
+
+export function RouterProvider( {
```

Comment:
What's our current fallback in case someone tries to access a route that doesn't exist?

Reply:
It seems to land always in the home page in trunk. Feels a bit weird.

Reply:
At least it's not a SEO problem, but still, we might want to implement a 404 page at some point.